p2p/smc_service.py

```diff
@@ -0,0 +1,153 @@
+import asyncio
+import collections
+from typing import (
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Set,
+)
+
+from evm.rlp.headers import (
+    CollationHeader,
+)
+
+from evm.exceptions import (
+    ValidationError,
+)
+from eth_typing import (
+    Address,
+    Hash32,
+)
+from eth_utils import (
+    to_checksum_address,
+)
+from eth_keys import (
+    datatypes,
+)
+
+from p2p.service import (
+    BaseService,
+)
+from p2p.cancel_token import (
+    CancelToken,
+)
+
+from web3 import (
+    Web3,
+)
+
+from sharding.handler.smc_handler import (
+    SMCHandler,
+)
+from sharding.handler.log_handler import (
+    LogHandler,
+)
+from sharding.handler.shard_tracker import (
+    ShardTracker,
+)
+from sharding.contracts.utils.config import (
+    get_sharding_config,
+)
+
+
+class SMCService(BaseService):
+    w3: Web3 = None
+    config = get_sharding_config()
+
+    def __init__(self,
+                 smc_address: Address,
+                 shard_ids: List[int]=None,
+                 private_key: Optional[datatypes.PrivateKey]=None,
+                 token: CancelToken=None) -> None:
+        if self.w3 is None:
+            raise ValueError("Web3 connection has not been configured yet")
+
+        super().__init__(token)
+        self.smc_address = smc_address
+        self.private_key = private_key
+
+        self._smc_handler = self.w3.eth.contract(
+            address=self.smc_address,
+            ContractFactoryClass=SMCHandler,
+            config=self.config,
+        )
+        self._log_handler = LogHandler(self.w3, self.config["PERIOD_LENGTH"])
+
+        self._shard_trackers: Dict[int, ShardTracker] = {}
+        self._added_header_queues: DefaultDict[
+            int,
+            List[asyncio.Queue[CollationHeader]]
+        ] = collections.defaultdict(list)
+        for shard_id in shard_ids or []:
+            self.start_watching_shard(shard_id)
+
+        self._processed_header_hashes: Set[Hash32] = set()
+
+    @property
+    def add_header_gas_price(self) -> int:
+        return 11 * 1000 * 1000 * 1000
+
+    def start_watching_shard(self, shard_id: int) -> asyncio.Queue[CollationHeader]:
+        if shard_id in self._shard_trackers:
+            raise ValueError("Already watching shard {}".format(shard_id))
+
+        self.logger.debug("Start watching shard {}".format(shard_id))
+        shard_tracker = ShardTracker(self.config, shard_id, self._log_handler, self.smc_address)
+        queue: asyncio.Queue[CollationHeader] = asyncio.Queue()
+        self._shard_trackers[shard_id] = shard_tracker
```

Comment:
Instead of two data structures, what about using `collections.namedtuple` for these two fields.